src/utils/generator.js

```diff
@@ -70,22 +100,141 @@ export type GeneratorEntry = {
+
+  declared: void | AbstractValue;
+  args: Array<Value>;
+  // If we're just trying to add roots for the serializer to notice, we don't need a buildNode.
+  buildNode: void | GeneratorBuildNodeFunction;
+  dependencies: void | Array<Generator>;
+  isPure: void | boolean;
+
+  visit(callbacks: VisitEntryCallbacks) {
+    if (this.isPure && this.declared && callbacks.canSkip(this.declared)) {
+      callbacks.recordDelayedEntry(this.containingGenerator, this);
+    } else {
+      if (this.declared) callbacks.recordDeclaration(this.declared);
+      callbacks.visitValues(this.args);
+      if (this.dependencies)
+        for (let dependency of this.dependencies) callbacks.visitGenerator(dependency, this.containingGenerator);
+    }
+  }
+
+  serialize(context: SerializationContext) {
+    if (!this.isPure || !this.declared || !context.canOmit(this.declared)) {
+      let nodes = this.args.map((boundArg, i) => context.serializeValue(boundArg));
+      if (this.buildNode) {
+        let node = this.buildNode(nodes, context);
+        if (node.type === "BlockStatement") {
+          let block: BabelNodeBlockStatement = (node: any);
+          let statements = block.body;
+          if (statements.length === 0) return;
+          if (statements.length === 1) {
+            node = statements[0];
+          }
+        }
+        context.emit(node);
+      }
+      if (this.declared !== undefined) context.declare(this.declared);
+    }
+  }
+}
+
+type ModifiedPropertyEntryArgs = {|
+  propertyBinding: PropertyBinding,
+  // TODO make these mutable?
+  newValue: void | Descriptor,
+  oldValue: void | Descriptor,
 |};
 
+class ModifiedPropertyEntry extends GeneratorEntry {
+  constructor(generator: Generator, args: ModifiedPropertyEntryArgs) {
+    super(generator);
+    Object.assign(this, args);
+  }
+
+  propertyBinding: PropertyBinding;
+  // TODO make these mutable?
+  newValue: void | Descriptor;
+  oldValue: void | Descriptor;
+
+  serialize(context: SerializationContext) {
+    let desc = this.propertyBinding.descriptor;
+    invariant(desc === this.newValue);
+    context.emitPropertyModification(this.propertyBinding);
+  }
+
+  visit(context: VisitEntryCallbacks) {
+    let desc = this.propertyBinding.descriptor;
+    invariant(desc === this.newValue);
+    context.visitObjectProperty(this.propertyBinding);
+  }
+}
+
+type ModifiedBindingEntryArgs = {|
+  modifiedBinding: Binding,
+  newValue: BindingEntry,
+  oldValue: void | Value,
+|};
+
+class ModifiedBindingEntry extends GeneratorEntry {
+  constructor(generator: Generator, args: ModifiedBindingEntryArgs) {
+    super(generator);
+    Object.assign(this, args);
+  }
+
+  containingGenerator: Generator;
+  modifiedBinding: Binding;
+  newValue: BindingEntry;
+  oldValue: void | Value;
+  residualFunctionBinding: ?ResidualFunctionBinding;
+
+  serialize(context: SerializationContext) {
+    let residualFunctionBinding = this.residualFunctionBinding;
+    if (!residualFunctionBinding) return;
```

Comment:
How can that be? How can there be an attempt to serialize if it wasn't visited? (And visiting seems to guarantee setting `this.residualFunctionBinding` to a defined value.